scipy/stats/_distn_infrastructure.py

```diff
@@ -3836,3 +3884,423 @@ def get_distribution_names(namespace_pairs, rv_base_class):
+    value between 0 and 1. In such a case - when the domain of the distribution
+    with respect to a parameter is finite - we are not required to specify
+    bounds for the parameter.
+
+    >>> shape_bounds = {'n': (0, 30)}  # omit parameter p using a `dict`
+    >>> res2 = stats.fit(dist, data, shape_bounds)
+    >>> np.testing.assert_allclose(res.params, res2.params)
+    True
+
+    If we wish to force the distribution to be fit with *n* fixed at 6, we can
+    set both the lower and upper bounds on *n* to 6. Note, however, that the
+    value of the objective function being optimized is worse (higher) in this
+    case.
+
+    >>> shape_bounds = {'n': (6, 6)}  # fix parameter `n`
+    >>> res3 = stats.fit(dist, data, shape_bounds)
+    >>> res3.params
+    FitShapes(n=6.0, p=0.5486556076755706, loc=0.0)  # may vary
+    >>> res3.objective_val > res.objective_val
+    True
+
+    Note that any `optimizer` parameter allows us to control the optimizer
+    used to perform the fitting. The default optimizer is
+    `scipy.optimize.differential_evolution` with its own default settings,
+    but we can easily change these settings by creating our own optimizer
+    with different defaults.
+
+    >>> from scipy.optimize import differential_evolution
+    >>> def optimizer(fun, bounds, *, integrality):
+    ...     return differential_evolution(fun, bounds, strategy='best2bin',
+    ...                                   seed=rng, integrality=integrality)
+    >>> shape_bounds = [(0, 30), (0, 1)]
+    >>> res4 = stats.fit(dist, data, shape_bounds, optimizer=optimizer)
+    >>> res4.params
+    FitShapes(n=5.0, p=0.5032774713044523, loc=0.0)  # may vary
+
+    """
+
+    # --- Input Validation / Standardization --- #
+
+    # distribution input validation and information collection
+    if getattr(dist, "pdf", False):  # can't use isinstance for types
+        loc_bounds = loc_bounds or (0, 0)
+        scale_bounds = scale_bounds or (1, 1)
+        loc_scale_integrality = [False, False]
+        discrete = False
+    elif getattr(dist, "pmf", False):
+        loc_bounds = loc_bounds or (0, 0)
+        loc_scale_integrality = [True]
+        discrete = True
+    else:
+        message = ("`dist` must be an instance of `rv_continuous` "
+                   "or `rv_discrete.`")
+        raise ValueError(message)
+
+    # data input validation
+    data = np.asarray(data)
+    if data.ndim != 1:
+        message = "`data` must be exactly one-dimensional."
+        raise ValueError(message)
+    if not (np.issubdtype(data.dtype, np.number)
+            and np.all(np.isfinite(data))):
+        message = "All elements of `data` must be finite numbers."
+        raise ValueError(message)
+
+     # shape_bounds input validation and information collection
+    shape_info = dist._get_shape_info()
+
+    n_shapes = len(shape_info)
+    if shape_bounds is None or np.array(shape_bounds).size == 0:
+        shape_bounds = {}
+
```

Comment:
`shape_bounds` can be either a list of tuples or a dict. If a dict, standardize to array form, filling in any blanks with `shape_info` and warning if there are unrecognized shape names. If an array, all shape bounds are required, so check the shape.